store/cachekv/internal/memiterator.go

```diff
@@ -0,0 +1,137 @@
+package internal
+
+import (
+	"bytes"
+	"errors"
+
+	"github.com/cosmos/cosmos-sdk/store/types"
+	"github.com/tidwall/btree"
+)
+
+var _ types.Iterator = (*memIterator)(nil)
+
+// memIterator iterates over iterKVCache items.
+// if key is nil, means it was deleted.
+// Implements Iterator.
+type memIterator struct {
+	iter btree.GenericIter[item]
+
+	start     []byte
+	end       []byte
+	ascending bool
+	lastKey   []byte
+	deleted   map[string]struct{}
+	valid     bool
+}
+
+func NewMemIterator(start, end []byte, items *BTree, deleted map[string]struct{}, ascending bool) *memIterator {
+	iter := items.tree.Iter()
+	var valid bool
+	if ascending {
+		if start != nil {
+			valid = iter.Seek(newKey(start))
+		} else {
+			valid = iter.First()
+		}
+	} else {
+		if end != nil {
+			valid = iter.Seek(newKey(end))
+			if !valid {
+				valid = iter.Last()
+			} else {
+				// end is exclusive
+				valid = iter.Prev()
+			}
+		} else {
+			valid = iter.Last()
+		}
+	}
+
+	mi := &memIterator{
+		iter:      iter,
+		start:     start,
+		end:       end,
+		ascending: ascending,
+		lastKey:   nil,
+		deleted:   deleted,
+		valid:     valid,
+	}
+
+	if mi.valid {
+		mi.valid = mi.keyInRange(mi.Key())
+	}
+
+	return mi
+}
+
+func (mi *memIterator) Domain() (start []byte, end []byte) {
+	return mi.start, mi.end
+}
+
+func (mi *memIterator) Close() error {
+	mi.iter.Release()
+	return nil
+}
+
+func (mi *memIterator) Error() error {
+	if !mi.Valid() {
+		return errors.New("invalid memIterator")
+	}
+	return nil
+}
+
+func (mi *memIterator) Valid() bool {
```

Comment:
Why are we defining public methods on a type that's used internally? A caller can just do `mi.valid` as an example.

Reply:
These are methods of interface `types.Iterator`.